Expert Watermelon Guppy

High

# Batch sending tokens upon a clawback can result in failed clawbacks

### Summary

Batch sending tokens upon a clawback can result in failed clawbacks

### Root Cause

Upon having a clawback, we have the following piece of code:
```go
	spendable := m.bk.SpendableCoins(ctx, vestingAccAddr)
	toXfer := coinsMin(toClawBack, spendable)
	if toXfer.IsAllPositive() {
		err = m.bk.SendCoins(ctx, vestingAccAddr, funderAddr, toXfer)
		if err != nil {
			return nil, err // shouldn't happen, given spendable check
		}
	}
```
This code is problematic as we are conducting a batch transfer of the tokens. If any of them fails for any reasons, such as a blacklist, we will simply return an error and the clawback functionality will be unusable, the funder can not get his tokens back which is essentially a loss of funds for him.

The `spendable` value only considers the balances but not the fact that the transfer can fail, for reasons such as a blacklist. Note that the vesting can be done for any token which automatically includes tokens that have such functionality.


### Internal Pre-conditions

1. Transfer failing due to a reason such as a blacklist

### External Pre-conditions

_No external pre-conditions_

### Attack Path

1. Bob wants to clawback his tokens from Alice, 100 tokens of token `A` and 100 tokens of token `B`
2. Alice is blacklisted for token `B`
3. The batch transfer will return an error upon the transfer of token `B`, Bob can not clawback his tokens

### Impact

Loss of funds.

### PoC

_No response_

### Mitigation

Do not use a batch send. Instead, send each coin separately.